pkg/resource/deploy/snapshot.go

```diff
@@ -67,6 +67,46 @@ func NewSnapshot(manifest Manifest, secretsManager secrets.Manager,
 	}
 }
 
+// NormalizeURNReferences fixes up all URN references in a snapshot to use the new URNs instead of potentially-aliased
+// URNs.  This will affect resources that are "old", and which would be expected to be updated to refer to the new names
+// later in the deployment.  But until they are, we still want to ensure that any serialization of the snapshot uses URN
+// references which do not need to be indirected through any alias lookups, and which instead refer directly to the URN
+// of a resource in the resources map.
+//
+// Note: This method modifies the snapshot (and resource.States in the snapshot) in-place.
+func (snap *Snapshot) NormalizeURNReferences() error {
+	if snap != nil {
+		aliased := make(map[resource.URN]resource.URN)
+		fixUrn := func(urn resource.URN) resource.URN {
+			if newUrn, has := aliased[urn]; has {
+				return newUrn
+			}
+			return urn
+		}
+		for _, state := range snap.Resources {
+			// Fix up any references to URNs
```

Comment:
This also needs to rewrite provider references.

Reply:
Done.